ops/jujucontext.py

```diff
@@ -0,0 +1,229 @@
+    """The dispatch path in the format of 'actions/do-something'.
+
+    For example 'hooks/workload-pebble-ready' (from JUJU_DISPATCH_PATH).
+    """
+
+    model_name: str = ''
+    """The name of the model.
+
+    For example 'foo' (from JUJU_MODEL_NAME).
+    """
+
+    model_uuid: str = ''
+    """The uuid of the model.
+
+    For example 'cdac5656-2423-4388-8f30-41854b4cca7d' (from JUJU_MODEL_UUID).
+    """
+
+    notice_id: Optional[str] = None
+    """The ID of the notice.
+
+    For example '1', (from JUJU_NOTICE_ID).
+    """
+
+    notice_key: Optional[str] = None
+    """The key of the notice.
+
+    For example 'example.com/a', (from JUJU_NOTICE_KEY).
+    """
+
+    notice_type: Optional[str] = None
+    """The type of the notice.
+
+    For example 'custom' (from JUJU_NOTICE_TYPE).
+    """
+
+    pebble_check_name: Optional[str] = None
+    """The name of the pebble check.
+
+    For example 'http-check' (from JUJU_PEBBLE_CHECK_NAME).
+    """
+
+    relation_departing_unit_name: Optional[str] = None
+    """The unit that is departing a relation.
+
+    For example 'remote/42' (from JUJU_DEPARTING_UNIT).
+    """
+
+    relation_name: Optional[str] = None
+    """The name of the relation.
+
+    For example 'database' (from JUJU_RELATION).
+    """
+
+    relation_id: Optional[int] = None
+    """The id of the relation.
+
+    For example 1 (integer) if the original environment variable's value is 'database:1'
+    (from JUJU_RELATION_ID).
+    """
+
+    remote_app_name: Optional[str] = None
+    """The name of the remote app.
+
+    For example 'remoteapp1' (from JUJU_REMOTE_APP).
+    """
+
+    remote_unit_name: Optional[str] = None
+    """The name of the remote unit.
+
+    For example 'remoteapp1/0' (from JUJU_REMOTE_UNIT).
+    """
+
+    secret_id: Optional[str] = None
+    """The ID of the secret.
+
+    For example 'secret:dcc7aa9c-7202-4da6-8d5f-0fbbaa4e1a41' (from JUJU_SECRET_ID).
+    """
+
+    secret_label: Optional[str] = None
+    """The label of the secret.
+
+    For example 'db-password' (from JUJU_SECRET_LABEL).
+    """
+
+    secret_revision: Optional[int] = None
+    """The revision of the secret.
+
+    For example 42 (integer) (from JUJU_SECRET_REVISION).
+    """
+
+    storage_name: Optional[str] = None
+    """The storage name.
+
+    For example 'my-storage' if the original environment variable's value is 'my-storage/1'
+    (from JUJU_STORAGE_ID).
+    """
+
+    unit_name: str = ''
+    """The name of the unit.
+
+    For example 'myapp/0' (from JUJU_UNIT_NAME).
+    """
+
+    version: JujuVersion = dataclasses.field(default_factory=lambda: JujuVersion('0.0.0'))
+    """The version of Juju.
+
+    For example '3.4.0' (from JUJU_VERSION).
+    """
+
+    workload_name: Optional[str] = None
+    """The name of the workload.
+
+    For example 'workload' (from JUJU_WORKLOAD_NAME).
+    """
+
+    @classmethod
+    def from_dict(cls, env: Mapping[str, Any]) -> '_JujuContext':
```

Comment:
Design question:

The half-baked idea I imagined was that `from_dict` returns a `Union` of types, and the caller always uses a [match](https://peps.python.org/pep-0636/) expression on the result. It would imply py3.10 (ubuntu 22.04) which isn't great, but I wonder (a) if it's a feasible design goal in the first place, and (b) if we should plan the design today with `match` in mind.

I.e. it would be great if this new class brought significant new added value over poking into `os.environ` directly.

Reply:
Example: https://stackoverflow.com/a/71519690

Reply:
If I understand correctly, there is already some similar logic in `main.py`, specifically, in [`_get_event_args`](https://github.com/canonical/operator/blob/main/ops/main.py#L154), where it tries to match the type of the event then pull related environment variables.

I think it would be great if we could somehow integrate it with JujuContext, maybe in the future we can build on top of this PR and implement that.

Regarding the scope of this PR, we had a short discussion today in the daily and we think this PR serves two purposes: 1, parse all ENV vars from a single place (previously it was scattered across `main.py` and `framework.py`); and 2, provide some unified object so that others can build on top of it if they want to implement some experiments.

So, for now, I'm sorry that there is no implementation of the event type matching thingy and we are going to merge this `_JujuContext`. Thanks!

Reply:
> The half-baked idea I imagined was that `from_dict` returns a `Union` of types, and the caller always uses a [match](https://peps.python.org/pep-0636/) expression on the result.

If the intention is to find out which event Juju has emitted, then it's much simpler to just look at that directly rather than trying to match against which context variables have been set. If the intention is to get back a collection of types (e.g. which fields are set are the same with different events - start and stop would be setting the same) then I'm not sure I really see the use-case.

I don't think it's good to have a hierarchy of context objects when there's already a 1:1 mapping of what that would be in the hierarchy of event classes. A `SecretChangedJujuContext` would only be useful for ` SecretChangedEvent`, for example.

As @IronCore864 mentioned, `_get_event_args` provides the "given a context, what are the arguments needed to create an event object". It would be simple enough for someone to subclass `_JujuContext` to add a `to_event` method (or whatever), if they wanted to build an "alt ops" package.

What would be most useful in that sort of situation, I believe, is being able to create `SecretChangedEvent` (and on on) objects without opting in to the whole framework/handle system (and registering, in particular). My feeling is that's what we could provide (not this cycle) to continue making it easier for people to explore alternative ways to do event handling without re-implementing everything.

> I.e. it would be great if this new class brought significant new added value over poking into `os.environ` directly.

"significant" is subjective, of course, but I think there is value in not just collecting all the uses together as @IronCore864 mentioned, but also doing all the conversion so that you have a bunch of Python objects rather than a bunch of strings. With this, plus the similar hook tool wrapping in the model/model backend (the border of those could be a bit cleaner) that's the main interface with Juju (other than Pebble).